invokeai/app/services/image_record_storage.py

```diff
@@ -0,0 +1,353 @@
+
+        def __init__(self, message="Image record not found"):
+            super().__init__(message)
+
+    class ImageRecordSaveException(Exception):
+        """Raised when an image record cannot be saved."""
+
+        def __init__(self, message="Image record not saved"):
+            super().__init__(message)
+
+    class ImageRecordDeleteException(Exception):
+        """Raised when an image record cannot be deleted."""
+
+        def __init__(self, message="Image record not deleted"):
+            super().__init__(message)
+
+    @abstractmethod
+    def get(self, image_type: ImageType, image_name: str) -> ImageRecord:
+        """Gets an image record."""
+        pass
+
+    @abstractmethod
+    def get_many(
+        self,
+        image_type: ImageType,
+        image_category: ImageCategory,
+        page: int = 0,
+        per_page: int = 10,
+    ) -> PaginatedResults[ImageRecord]:
+        """Gets a page of image records."""
+        pass
+
+    @abstractmethod
+    def delete(self, image_type: ImageType, image_name: str) -> None:
+        """Deletes an image record."""
+        pass
+
+    @abstractmethod
+    def save(
+        self,
+        image_name: str,
+        image_type: ImageType,
+        image_category: ImageCategory,
+        session_id: Optional[str],
+        node_id: Optional[str],
+        metadata: Optional[ImageMetadata],
+        created_at: str = datetime.datetime.utcnow().isoformat(),
+    ) -> None:
+        """Saves an image record."""
+        pass
+
+
+class SqliteImageRecordStorage(ImageRecordStorageBase):
+    _filename: str
+    _conn: sqlite3.Connection
+    _cursor: sqlite3.Cursor
+    _lock: threading.Lock
+
+    def __init__(self, filename: str) -> None:
+        super().__init__()
+
+        self._filename = filename
+        self._conn = sqlite3.connect(filename, check_same_thread=False)
+        # Enable row factory to get rows as dictionaries (must be done before making the cursor!)
+        self._conn.row_factory = sqlite3.Row
+        self._cursor = self._conn.cursor()
+        self._lock = threading.Lock()
+
+        try:
+            self._lock.acquire()
+            # Enable foreign keys
+            self._conn.execute("PRAGMA foreign_keys = ON;")
+            self._create_tables()
+            self._conn.commit()
+        finally:
+            self._lock.release()
+
+    def _create_tables(self) -> None:
+        """Creates the tables for the `images` database."""
+
+        # Create the `images` table.
+        self._cursor.execute(
+            f"""--sql
+            CREATE TABLE IF NOT EXISTS images (
+                id TEXT PRIMARY KEY,
+                image_type TEXT, -- non-nullable via foreign key constraint
+                image_category TEXT, -- non-nullable via foreign key constraint
+                session_id TEXT, -- nullable
+                node_id TEXT, -- nullable
+                metadata TEXT, -- nullable
+                created_at TEXT NOT NULL,
```

Comment:
I think its better to rely on the DB to populate this (versus defining it in code). Brief googling says that you may be bale to do something like `created_at DATETIME DEFAULT CURRENT_TIMESTAMP`. Should we also include updated_at?